designs/2021-fixable-disable-directives/design.md

```diff
@@ -0,0 +1,112 @@
+
+This RFC proposes making three changes:
+
+- In the patched `options.fix`, consider a problem's meta type to be `"meta"` if its `ruleId` is `null`
+- Pass the `SourceCode` being linted as a parameter to `applyDisableDirectives`
+- Add a `fix` method to the unused directive problems returned by `applyDirectives` that uses the `SourceCode` to compute:
+  - For comments that are the only non-whitespace on their line, delete that line
+  - Otherwise, delete just the comment and any now-unnecessary surrounding whitespace
+
+### Fix Behavior Examples
+
+```diff
+- /* eslint-disable */
+```
+
+```diff
+- // eslint-disable-next-line -- related explanation
+```
+
+```diff
+- before /* eslint-disable-next-line -- related explanation */ after
++ before after
+```
+
+```diff
+- before // eslint-disable-next-line -- related explanation
++ before
+```
+
+```diff
+  // before
+- // eslint-disable-next-line
+  // after
+```
+
+Multiline block comments are already not allowed to be disable directives. [eslint/eslint#10334](https://github.com/eslint/eslint/issues/10334)
+
+## Documentation
+
+- [Command Line Interface > Fixing Problems](https://eslint.org/docs/user-guide/command-line-interface#fixing-problems)'s `--fix` documentation should mention the added directives fixing and the new `--fix-type`
+- [Rules > Report Unused ESLint Disable Comments](https://eslint.org/docs/user-guide/configuring/rules#report-unused-eslint-disable-comments) should link to that documentation
+
+## Drawbacks
+
+Like any new feature, this flag will slightly increase the complexity and maintenance costs of ESLint core.
+
+This RFC's implementation would lock in the name for a new `--fix-type` even though we only have one concrete use case for it.
+
+## Backwards Compatibility Analysis
+
+It is unlikely but not impossible that some `--fix` dependant users would rely on unused disable directives remaining in code.
+
+Otherwise, this change is additive in behavior.
+
+## Alternatives
+
+- Applying directive fixes after the `verifyAndFix` passes
+  - Not ideal: deleting a comment could introduce new rule violations that would warrant running >=1 other pass
+- Writing an external tool to apply these options
+  - Not ideal: the internal implementation is simple; external would have to deal with variant formatters, etc and suffer from the same introduced rule violations
+
+## Open Questions
+
+- Is `"meta"` a good name for the fix type?
+  - I'd considered `"directive"`, but that felt like it could be too specific to be extended for other meta/configuration in the future.
+
+## Help Needed
+
+I'm looking forward to implementing this if approved! 🙌
+
+## Frequently Asked Questions
+
+> Will these be autofixed by default?
+
+Yes: problems reported by directive usage checking are joined with remaining rule violation problems in a single array.
+This should allow directive fixing to seamlessly act similarly to rule fixing.
```

Comment:
https://github.com/eslint/rfcs/pull/78#issuecomment-817466031

> If this won’t be autofixed by default (it should be) I would like a way to configure it in eslintrc (and a shared config) so that it is.

@ljharb added an FAQ entry here for my interpretation of the question. Does this answer what you're looking for?

Reply:
awesome, yes thanks, that's what i'd hope for.